test/lib/storageUtils.spec.js

```diff
@@ -174,3 +176,106 @@ describe('value functions', ()=>{
+    let cookies = [];
+    let getStub, delStub, setStub;
+
+    /**
+     * Return a mock function for setting cookies
+     * @param levels Levels of TLD.  For ex, '.com' is 1, and '.co.za' is 2.
+     * @returns {function(*, *, *, *): void}
+     */
+    function genSetCookie(levels) {
+        return (name, value, expires, domain, path) => {
+            const domainParts = domain.split('.');
+            if (levels > 0 && domainParts.length > levels) {
+                // find existing ones first
+                const element = cookies.find(e => e.name === name && e.domain === domain && e.path === path);
+
+                if (element) {
+                    element.value = value;
+                }
+                else {
+                    const newElement = {
+                        name: name,
+                        value: value,
+                        domain: domain,
+                        path: path
+                    };
+                    cookies.push(newElement);
+                }
+            }
+        }
+    }
+
+    function mockGetCookie(name) {
+        const c = cookies.find(element => element.name === name);
+        return c ? c.value : undefined;
+    }
+
+    function mockDelCookie(name, domain, path) {
+        for (let i = 0; i < cookies.length; ++i) {
+            const element = cookies[i];
+            if (element.name === name && element.domain === domain && element.path === path) {
+                cookies.splice(i, 1);
+                break;
+            }
+        }
+    }
+
+    beforeEach(()=>{
+       cookies = [];
+       getStub = sinon.stub(cookieUtils, 'getCookie').callsFake(mockGetCookie);
+       delStub = sinon.stub(cookieUtils, 'delCookie').callsFake(mockDelCookie);
+    });
+
+    afterEach(()=>{
+        cookies = [];
+        if (setStub) setStub.restore();
+        getStub.restore();
+        delStub.restore();
+    });
+
+    it('1 tld with subdomain', ()=>{
+        setStub = sinon.stub(cookieUtils, 'setCookie').callsFake(genSetCookie(1));
+        const domain = storage.extractDomain('www.example.com');
+        expect(domain).to.equal('example.com');
+        expect(cookies).to.have.lengthOf(0);
+    });
+
+    it('1 tld without subdomain', ()=>{
+        setStub = sinon.stub(cookieUtils, 'setCookie').callsFake(genSetCookie(1));
+        const domain = storage.extractDomain('example.com');
+        expect(domain).to.equal('example.com');
+        expect(cookies).to.have.lengthOf(0);
+    });
+
+    it('2 tld with subdomain', ()=>{
+        setStub = sinon.stub(cookieUtils, 'setCookie').callsFake(genSetCookie(2));
+        const domain = storage.extractDomain('www.example.uk.com');
+        expect(domain).to.equal('example.uk.com');
+        expect(cookies).to.have.lengthOf(0);
+    });
+
+    it('3 tld with subdomain', ()=>{
+        setStub = sinon.stub(cookieUtils, 'setCookie').callsFake(genSetCookie(3));
+        const domain = storage.extractDomain('a.b.c.kobe.jp');
```

Comment:
I understand this is a hard case. Just wondering if the best answer would be 'kobe.jp'  but we are settling for good enough here?

Reply:
jp domains have some pretty complex rules.  As I understand it, this is correct.  The examples come from [psl](https://www.npmjs.com/package/psl).  We could have just used psl or something similar without messing with the cookie test, but it more than double the size of the final script.  

Anyway, the mock cookie methods aren't really validating domains.  They are just there to simulate different levels of op-level domains.